### PR TITLE
Add staging actuator to Coil:Heating:Electric:MultiStage

### LINKED ISSUE #9357: Multi-stage electric backup coil modeling
Issue overview
--------------
It is not uncommon that the multispeed heat pumps might come with a multi-stage back up coil. But there's no explicit capability in current E+ to allow modeling a multistage electric coil.  There're mainly two restrictions:
1. The `Coil:Heating:Electric:MultiStage` looks like a good fit to model a multi-stage electric coil, but it is not an option of [`Supplemental Heating Coil Object Type` for AirLoopHVAC:UnitarySystem parent object](https://bigladdersoftware.com/epx/docs/9-6/input-output-reference/group-unitary-equipment.html#field-supplemental-heating-coil-object-type). It is documented as "For the time being, this coil model can only be called by the parent object [AirLoopHVAC:UnitaryHeatPump:AirToAir:MultiSpeed.](https://bigladdersoftware.com/epx/docs/9-6/input-output-reference/group-unitary-equipment.html#airloophvacunitaryheatpumpairtoairmultispeed)" There're a few objects that are only able to be called by specific parent objects, but considering AirLoopHVAC:UnitarySystem is supposed to serve as an universal object that covers most functionalities, should we consider allowing Coil:Heating:Electric:MultiStage to be with AirLoopHVAC:UnitarySystem as well?
2. And there's no explicit method to control the backup coil staging. To model realistic hvac cycling, the heat pump coil is operating with an [on-off thermostat](https://github.com/NREL/EnergyPlus/pull/6385), and in this context, the current modeling flexibility is not sufficient to implement the staging of backup coil. A stage actuator or a schedule input is needed to allow EMS program for specific backup coil staging. Similar to the purpose of https://github.com/NREL/EnergyPlus/pull/8865.  

This capability is important for our GEB hvac modeling and analysis, so we want to get some feedback from E+ dev team on this. Let me know if you have any thought. Thanks!

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by yzhou601**
Pull request overview
---------------------
 - For enhancement requested by #9357 
 - This PR is to add an actuator for Coil:Heating:Electric:MultiStage object to offer flexibility to its staging logic with EMS.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by nrel-bot-2** [comment]
@yzhou601 @myoldmopar it has been 28 days since this pull request was last updated.

**Comment by rraustad** [comment]
Are these changes a result of findings from #9525?

**Comment by yzhou601** [comment]
> 

No, they're changes to introduce new actuator, not related to that PR.

**Comment by yzhou601** [comment]
@Myoldmopar This PR is ready for review. Test files show a good EMS control (see below the output). The high stage is kicked on when zone air temperature is lower than 1C below heating setpoint. It reports with "detailed" output frequency. There's a lag on EMS control but overall it looks good.

image

File: src/EnergyPlus/UnitarySystem.cc [hunk 1]
```diff
@@ -10980,15 +10984,16 @@ namespace UnitarySystems {
             }
         }
         if (this->m_SuppCoilExists) {
-            this->calcUnitarySuppHeatingSystem(state, FirstHVACIteration, this->m_SuppHeatPartLoadFrac, SuppCoilLoad);
+            this->calcUnitarySuppHeatingSystem(state, FirstHVACIteration, SuppCoilLoad);
```
Comment (yzhou601) [inline src/EnergyPlus/UnitarySystem.cc:10987]: The call all passes this->m_SuppHeatPartLoadFrac so I removed the argument.

File: src/EnergyPlus/UnitarySystem.cc [hunk 2]
```diff
@@ -7677,7 +7678,7 @@ namespace UnitarySystems {
                                       HeatCoilLoad,
                                       SupHeaterLoad,
                                       CompressorOn);
-
+        // Why need this second run?
```
Comment (yzhou601) [inline src/EnergyPlus/UnitarySystem.cc:7681]: The second call passes the same argument as previous call, I didn't see the point of running it second time. But I haven't tried removing it yet.

File: src/EnergyPlus/UnitarySystem.cc [hunk 3]
```diff
@@ -14247,95 +14312,21 @@ namespace UnitarySystems {
...
-                    // do nothing, coil is at set point (i.e., gas/elec/steam/user coil will try to hit set point)
-                } else if (PartLoadFrac > 0.0) {
-                    // do nothing, coil tried to hit set point (i.e., gas/elec/steam/user coil tried to hit set point but missed
-                } else if (NoLoadTempOut > (DesOutTemp - Acc)) {
-                    PartLoadFrac = 0.0; // outlet temp > set point, coil is not needed
-                } else {                // outlet temp too low, turn on coil
-
-                    // Get full load result
-                    PartLoadFrac = 1.0;
+                if (this->m_EMSOverrideSuppCoilSpeedNumOn) {
```
Comment (yzhou601) [inline src/EnergyPlus/UnitarySystem.cc:14315]: This part is just adding one if-else, not really changing too much in the else{} branching.

File: src/EnergyPlus/UnitarySystem.cc [hunk 4]
```diff
@@ -17891,6 +17964,13 @@ namespace UnitarySystems {
                                              state.dataUnitarySystems->unitarySys[sysNum].m_EMSOverrideCoilSpeedNumOn,
                                              state.dataUnitarySystems->unitarySys[sysNum].m_EMSOverrideCoilSpeedNumValue);
                         }
+                        SetupEMSActuator(state,
+                                         "Coil Speed Control",
```
Comment (yzhou601) [inline src/EnergyPlus/UnitarySystem.cc:17968]: Is it fine using the same actuator name as main multispeed heating/cooling coils?

**Review by yzhou601 — COMMENTED** [review]
I think there's only one remaining question on my end. There's no related multistage supplemental coil output variables for unitary system. Do we want to add some output variable eg. Unitary System Supplemental Coil Stage Level, runtime fraction? Or either those at coil object level? Currently there're only energy output variables for Coil:Heating:Electric:MultiStage.

File: src/EnergyPlus/UnitarySystem.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitarySystem.cc:17968]: As long as the actuator name / instance name / control type combination is unique then it should be fine.  If not, then we probably need a new name.

File: doc/ems-application-guide/src/ems-actuators/hvac-systems-001.tex [hunk 5]
```diff
@@ -146,6 +146,13 @@ \subsubsection{Unitary System Coil Speed Control}\label{unitary-system-coil-spee
...
 
+Another actuator called ``Coil Speed Control'' is available with control type ``Unitary System Supplemental Coil Stage Level''. This actuator is only available for the `AirLoopHVAC:UnitarySystem`.
+
+The EMS override coil speed value, “Unitary System Supplemental Coil Stage Level”, is a continuous number below maximum coil speed level allowed. With any EMS override coil speed value, if the input value is a integer, the speed level is set as the exact speed value input, with the cycling or speed ratio = 1.0. Otherwise, if the floating point part is greater than zero, the speed level is calculated as the closest integer greater than the EMS speed value, and the cycling/speed ratio is set as the floating point part of the EMS speed value. For example, if EMS overrides coil speed value = 1.2, the speed level number is set as 2 with a speed ratio at 0.2.
```
Comment (Myoldmopar) [inline doc/ems-application-guide/src/ems-actuators/hvac-systems-001.tex:151]: This is great.

File: src/EnergyPlus/UnitarySystem.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitarySystem.cc:7681]: Do you want to try?

File: src/EnergyPlus/UnitarySystem.cc [hunk 6]
```diff
@@ -167,9 +167,10 @@ namespace UnitarySystems {
...
           m_HeatingSpeedNum(0), m_SpeedNum(0), m_SuppHeatingCycRatio(0.0), m_SuppHeatingSpeedRatio(0.0), m_SuppHeatingSpeedNum(0),
-          m_EMSOverrideCoilSpeedNumOn(false), m_EMSOverrideCoilSpeedNumValue(0.0), m_CoilSpeedErrIdx(0), m_DehumidInducedHeatingDemandRate(0.0),
-          m_TotalAuxElecPower(0.0), m_HeatingAuxElecConsumption(0.0), m_CoolingAuxElecConsumption(0.0), m_ElecPower(0.0), m_ElecPowerConsumption(0.0),
-          m_LastMode(0), m_FirstPass(true), m_TotCoolEnergyRate(0.0), m_SensCoolEnergyRate(0.0), m_LatCoolEnergyRate(0.0), m_TotHeatEnergyRate(0.0),
+          m_EMSOverrideCoilSpeedNumOn(false), m_EMSOverrideCoilSpeedNumValue(0.0), m_EMSOverrideSuppCoilSpeedNumOn(false),
+          m_EMSOverrideSuppCoilSpeedNumValue(0.0), m_CoilSpeedErrIdx(0), m_DehumidInducedHeatingDemandRate(0.0), m_TotalAuxElecPower(0.0),
+          m_HeatingAuxElecConsumption(0.0), m_CoolingAuxElecConsumption(0.0), m_ElecPower(0.0), m_ElecPowerConsumption(0.0), m_LastMode(0),
+          m_FirstPass(true), m_TotCoolEnergyRate(0.0), m_SensCoolEnergyRate(0.0), m_LatCoolEnergyRate(0.0), m_TotHeatEnergyRate(0.0),
```
Comment (Myoldmopar) [inline src/EnergyPlus/UnitarySystem.cc:173]: On future code changes, if you can just initialize struct member variables right where they are declared, you won't have to set them in this initializer list.  Not necessary to change it now, just try to do that moving forward.

File: src/EnergyPlus/UnitarySystem.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitarySystem.cc:10987]: Great, no need to pass member variables when calling methods, :+1:

File: src/EnergyPlus/UnitarySystem.cc [hunk 7]
```diff
@@ -11776,6 +11802,45 @@ namespace UnitarySystems {
...
+            this->m_SuppHeatPartLoadFrac = this->m_SuppHeatingCycRatio;
+        } else {
+            this->m_SuppHeatingCycRatio = 1.0;
+            this->m_SuppHeatingSpeedRatio = this->m_EMSOverrideSuppCoilSpeedNumValue - floor(this->m_EMSOverrideSuppCoilSpeedNumValue);
+            if (useMaxedSpeed || this->m_SuppHeatingSpeedRatio == 0) {
+                this->m_SuppHeatingSpeedRatio = 1;
+            }
+            this->m_SuppHeatPartLoadFrac = this->m_SuppHeatingSpeedRatio;
+        }
+    }
```
Comment (Myoldmopar) [inline src/EnergyPlus/UnitarySystem.cc:11842]: This worker function seems very reasonable, and testable.  :+1:

File: src/EnergyPlus/UnitarySystem.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitarySystem.cc:14315]: OK, it's pretty hard to track on Github's diff view, but I think I'm grokking it.

File: testfiles/PythonPlugin_SingleFamilyHouse_TwoSpeed_MultiStageElectricSuppCoil.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from pyenergyplus.plugin import EnergyPlusPlugin
+
+
+class SuppCoilSpeedControl(EnergyPlusPlugin):
```
Comment (Myoldmopar) [inline testfiles/PythonPlugin_SingleFamilyHouse_TwoSpeed_MultiStageElectricSuppCoil.py:59]: :+1:

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc [hunk 9]
```diff
@@ -2534,6 +2534,528 @@ TEST_F(ZoneUnitarySysTest, UnitarySystemModel_MultiStageElecHeatCoil_Backup_Setp
...
+    state->dataLoopNodes->Node(3).MassFlowRateMax = thisSys->m_DesignMassFlowRate; // max at fan outlet so fan won't limit flow
+                                                                                   // Heating coil air outlet node = 2
+    state->dataScheduleMgr->Schedule(1).CurrentValue = 1.0;                        // Enable schedule without calling schedule manager
+
+    state->dataGlobal->BeginEnvrnFlag = true; // act as if simulation is beginning
+
+    // Backup Heating coil air outlet node = 2
+    state->dataLoopNodes->Node(2).TempSetPoint = 33.0;
+    thisSys->m_EMSOverrideSuppCoilSpeedNumOn = true;
+    thisSys->m_EMSOverrideSuppCoilSpeedNumValue = 1.4;
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:2992]: :+1:

**Review by Myoldmopar — APPROVED** [review]
This is really good stuff.  I don't see any issues in the code, unless you are wanting to try to remove the possibly redundant code.  The new EMS and PythonEMS examples are great, and the unit test seems to capture the handing of the new EMSOverride variables inside the unitary system's `simulate` method.

**Comment by Myoldmopar** [comment]
This seems ready to go in.  @rraustad do you want to take a pass over it since it is heavily in unitary system?

File: src/EnergyPlus/UnitarySystem.cc (hunk 2 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc:7681]: There is lot's of these excess calls in UnitarySystem. Soon as it is stable the clean up will begin.

File: src/EnergyPlus/UnitarySystem.cc (hunk 1 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc:10987]: Prior to refactor we passed the partloadratio.

File: src/EnergyPlus/UnitarySystem.cc [hunk 10]
```diff
@@ -11703,13 +11707,23 @@ namespace UnitarySystems {
...
-                if (this->m_SuppHeatCoilType_Num == DataHVACGlobals::Coil_HeatingElectric_MultiStage) {
-                    this->calcMultiStageSuppCoilStageByLoad(state, SuppHeatCoilLoad, FirstHVACIteration);
+                if (this->m_EMSOverrideSuppCoilSpeedNumOn) {
+                    if (SuppHeatCoilLoad > 0.0) {
+                        this->setEMSSuppCoilStagePLR(state);
+                    } else {
+                        this->m_SuppHeatingSpeedRatio = 0.0;
+                        this->m_SuppHeatingCycRatio = 0.0;
+                        this->m_SuppHeatPartLoadFrac = 0.0;
+                    }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:11722]: If EMS overrides the SpeedNum the model will only use that SpeedNum of there is a supp heating coil load?

File: src/EnergyPlus/UnitarySystem.cc [hunk 11]
```diff
@@ -11776,6 +11802,45 @@ namespace UnitarySystems {
...
     }
 
+    void UnitarySys::setEMSSuppCoilStagePLR(EnergyPlusData &state)
+    {
+        bool useMaxedSpeed = false;
+        int SpeedNumEMS = ceil(this->m_EMSOverrideSuppCoilSpeedNumValue);
+        if (SpeedNumEMS > this->m_NumOfSpeedSuppHeating) {
+            SpeedNumEMS = this->m_NumOfSpeedSuppHeating;
+            useMaxedSpeed = true;
+        }
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:11812]: \>= ?

File: src/EnergyPlus/UnitarySystem.cc (hunk 1 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc:10987]: [This](https://github.com/NREL/EnergyPlus/pull/9392#issuecomment-1195969564) figure looks reasonable and it is up to EMS to control any better than that. Regarding the new output variable for supp heater PLR, I'd say that certainly should be in the child and it couldn't hurt to add to the parent.  No need to do that here but it should get done.

**Comment by rraustad** [comment]
Just a few questions to answer and this should be good.

File: src/EnergyPlus/UnitarySystem.cc (hunk 10 above)
Comment (yzhou601) [reply src/EnergyPlus/UnitarySystem.cc:11722]: Yes I think so. the line 11716 is the condition that only set EMS values when there's a supp heat coil load.

File: src/EnergyPlus/UnitarySystem.cc (hunk 11 above)
Comment (yzhou601) [reply src/EnergyPlus/UnitarySystem.cc:11812]: The difference of > and >= is just whether to throw a warning when the EMS speed value is equal to coil speed number (useMaxedSpeed is used to throw warnings). My understanding is that we should allow this situation without throwing warnings.

**Comment by rraustad** [comment]
@Myoldmopar we are not going to get everything correct on a new model. This is working as expected for now so go ahead and merge.

**Comment by Myoldmopar** [comment]
Sounds great.  I merged the desuperheater branch, so let me just pull develop into here locally and test, then I'll merge.

**Comment by Myoldmopar** [comment]
Good to go, merging.  Thanks @yzhou601 and @rraustad

File: src/EnergyPlus/UnitarySystem.cc [hunk 12]
```diff
@@ -11776,6 +11802,45 @@ namespace UnitarySystems {
...
+        if (this->m_SuppHeatingSpeedNum == 1) {
+            this->m_SuppHeatingSpeedRatio = 0.0;
+            this->m_SuppHeatingCycRatio = this->m_EMSOverrideSuppCoilSpeedNumValue - floor(this->m_EMSOverrideSuppCoilSpeedNumValue);
+            if (useMaxedSpeed || this->m_SuppHeatingCycRatio == 0) {
+                this->m_SuppHeatingCycRatio = 1;
+            }
+            this->m_SuppHeatPartLoadFrac = this->m_SuppHeatingCycRatio;
+        } else {
+            this->m_SuppHeatingCycRatio = 1.0;
```
Comment (yzhou601) [inline src/EnergyPlus/UnitarySystem.cc:11835]: I found a bug during my testing that if the SpeedNumEMS is 0.0 then it went to this condition and set the m_SuppHeatingCycRatio and m_SuppHeatingSpeedRatio to be 1.0, which is not correct, they should be 0.0.

Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc:11835]: When should SpeedNum = 0? when the coil is off? Having a SpeedNum, CycRatio, and SpeedRatio is tricky.

**Pull request merged**